library/psa_crypto.c

```diff
@@ -4695,6 +4696,133 @@ static void psa_des_set_key_parity( uint8_t *data, size_t data_size )
+* - [SP800-56A] §5.6.1.2.2 or FIPS Publication 186-4: Digital Signature
+*   Standard (DSS) [FIPS186-4] §B.4.2 for elliptic curve keys.
+*/
+#if defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR) || \
+    defined(MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_PUBLIC_KEY) || \
+    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR) || \
+    defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_PUBLIC_KEY)
+static psa_status_t psa_generate_derived_ecc_key_weierstrass_helper(
+    psa_key_slot_t *slot,
+    size_t bits,
+    psa_key_derivation_operation_t *operation,
+    uint8_t **data
+    )
+{
+    unsigned key_out_of_range = 1;
+    mbedtls_mpi k;
+    mbedtls_mpi diff_N_2;
+    /* ret variable is initialized to 0 as it is
+       used only by MBEDTLS_MPI_CHK macro */
+    int ret = 0;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    mbedtls_mpi_init( &k );
+    mbedtls_mpi_init( &diff_N_2 );
+
+    psa_ecc_family_t curve = PSA_KEY_TYPE_ECC_GET_FAMILY(
+                                slot->attr.type );
+    mbedtls_ecp_group_id grp_id =
+        mbedtls_ecc_group_of_psa( curve, bits, 0 );
+
+    if( grp_id == MBEDTLS_ECP_DP_NONE )
+    {
+        status = PSA_ERROR_INVALID_ARGUMENT;
+        goto cleanup;
+    }
+
+    mbedtls_ecp_group ecp_group;
+    mbedtls_ecp_group_init( &ecp_group );
+
+    if( ( status = mbedtls_to_psa_error( mbedtls_ecp_group_load( &ecp_group, grp_id ) ) ) != 0 )
+        goto cleanup;
+
+    /* N is the boundary of the private key domain (ecp_group.N). */
+    /* Let m be the bit size of N. */
+    size_t m = ecp_group.nbits;
+
+    size_t m_bytes = PSA_BITS_TO_BYTES( m );
+
+    /* Note: This function is always called with *data == NULL and it
+     * allocates memory for the data buffer. */
+    *data = mbedtls_calloc( 1, m_bytes );
+    if( *data == NULL )
+    {
+        status = PSA_ERROR_INSUFFICIENT_MEMORY;
+        goto cleanup;
+    }
+
+    while ( key_out_of_range )
+    {
+        /* 1. Draw a byte string of length ceiling(m/8) bytes. */
+        if ( ( status = psa_key_derivation_output_bytes( operation, *data, m_bytes ) ) != 0 )
+            goto cleanup;
+
+        /* 2. If m is not a multiple of 8 */
+        if (m % 8)
+        {
+            /* Set the most significant
+            * (8 * ceiling(m/8) - m) bits of the first byte in
+            * the string to zero.
+            */
+            uint8_t clear_bit_mask = (1 << (m % 8)) - 1;
+            *data[0] &= clear_bit_mask;
+        }
+
+        /* 3. Convert the string to integer k by decoding it as a
+        *    big-endian byte string.
+        */
+        MBEDTLS_MPI_CHK(mbedtls_mpi_read_binary( &k, *data, m_bytes));
+
+        /* 4. If k > N - 2, discard the result and return to step 1.
+        *    Result of comparison is returned. When it indicates error
+        *    then this fuction is called again.
+        */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_int( &diff_N_2, &ecp_group.N, 2) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_grow( &k, diff_N_2.n ) );
```

Comment:
Note: with the current implementation of `mbedtls_mpi_read_binary()`, this is not necessary. However, that's currently not documented. I think it would be worth adding to the internal documentation of `mbedtls_mpi_read_binary()` (that is, in `bignum.c`, but the comment above the function, to make it clear it's part of the contract) what the number of limbs of the result is, so we can start relying on it.

Reply:
Should I remove `MBEDTLS_MPI_CHK( mbedtls_mpi_grow( &k, diff_N_2.n ) );` ?

Reply:
I think so, but before doing that we need to make it part of the (internal) contract of `mbedtls_mpi_read_binary()` by changing the comment in `bignum.c` to something like:
```
 * Import X from unsigned binary data, little endian
 *
 * This function is guaranteed to return an MPI with exactly the necessary
 * number of limbs (in particular, it does not skip 0s in the input).
```

Reply:
Oh, yes we should! IIRC there's library code that depends on that for constant-timeness, and there's definitely test code that depends on that for testing what it's supposed to test (operations on `mbedtls_mpi` objects with leading zero limbs).

Reply:
Well, more precisely, there's code that depends on `read_binary` keeping leading zeroes in the result. I'm not sure there's anything that depends on it shrinking the output object if it's already larger than it needs to be.

Reply:
Seem that the same comment can be also added to `mbedtls_mpi_read_binary_le()`?

Reply:
> I'm not sure there's anything that depends on it shrinking the output object if it's already larger than it needs to be.

This PR does. (Even before the last commit, I mean.) `mbedtls_mpi_lt_mpi_ct()` only works when its two arguments have the same allocated size. (So, actually, if we didn't want to rely on `read_binary()` doing what we want, we should have called `mbedtls_mpi_resize_clear()` not `mbedtls_mpi_grow()`.